eth/gasprice/gasprice.go

```diff
@@ -17,212 +17,152 @@
+func NewGasPriceOracle(backend ethapi.Backend, params GpoParams) *GasPriceOracle {
+	blocks := params.GpoBlocks
+	if blocks < 1 {
+		blocks = 1
+	}
+	percent := params.GpoPercentile
+	if percent < 0 {
+		percent = 0
 	}
-	minbase := new(big.Int).Mul(minprice, big.NewInt(100))
-	if params.GpobaseCorrectionFactor > 0 {
-		minbase = minbase.Div(minbase, big.NewInt(int64(params.GpobaseCorrectionFactor)))
+	if percent > 100 {
+		percent = 100
 	}
 	return &GasPriceOracle{
-		chain:    chain,
-		db:       db,
-		evmux:    evmux,
-		params:   params,
-		blocks:   make(map[uint64]*blockPriceInfo),
-		minBase:  minbase,
-		minPrice: minprice,
-		lastBase: minprice,
+		backend:     backend,
+		lastPrice:   params.GpoDefault,
+		checkBlocks: blocks,
+		minBlocks:   (blocks + 1) / 2,
+		maxBlocks:   blocks * 5,
+		percentile:  percent,
 	}
 }
 
-func (gpo *GasPriceOracle) init() {
-	gpo.initOnce.Do(func() {
-		gpo.processPastBlocks()
-		go gpo.listenLoop()
-	})
-}
-
-func (self *GasPriceOracle) processPastBlocks() {
-	last := int64(-1)
-	cblock := self.chain.CurrentBlock()
-	if cblock != nil {
-		last = int64(cblock.NumberU64())
-	}
-	first := int64(0)
-	if last > gpoProcessPastBlocks {
-		first = last - gpoProcessPastBlocks
-	}
-	self.firstProcessed = uint64(first)
-	for i := first; i <= last; i++ {
-		block := self.chain.GetBlockByNumber(uint64(i))
-		if block != nil {
-			self.processBlock(block)
+// SuggestPrice returns the recommended gas price.
+func (self *GasPriceOracle) SuggestPrice(ctx context.Context) (*big.Int, error) {
+	self.cacheLock.RLock()
+	lastHead := self.lastHead
+	lastPrice := self.lastPrice
+	self.cacheLock.RUnlock()
+
+	head, _ := self.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+	headHash := head.Hash()
+	if headHash == lastHead {
+		return lastPrice, nil
+	}
+
+	self.fetchLock.Lock()
+	defer self.fetchLock.Unlock()
+
+	// try checking the cache again, maybe the last fetch fetched what we need
+	self.cacheLock.RLock()
+	lastHead = self.lastHead
+	lastPrice = self.lastPrice
+	self.cacheLock.RUnlock()
+	if headHash == lastHead {
+		return lastPrice, nil
+	}
+
+	blockNum := head.Number.Uint64()
+	chn := make(chan lpResult, self.checkBlocks)
+	sent := 0
+	exp := 0
+	var lps bigIntArray
+	for sent < self.checkBlocks && blockNum > 0 {
+		go self.getBlockPrices(ctx, blockNum, chn)
+		sent++
+		exp++
+		blockNum--
+	}
+	maxEmpty := self.checkBlocks - self.minBlocks
+	for exp > 0 {
+		res := <-chn
+		if res.err != nil {
+			return lastPrice, res.err
 		}
-	}
-
-}
-
-func (self *GasPriceOracle) listenLoop() {
-	events := self.evmux.Subscribe(core.ChainEvent{}, core.ChainSplitEvent{})
-	defer events.Unsubscribe()
-
-	for event := range events.Chan() {
-		switch event := event.Data.(type) {
-		case core.ChainEvent:
-			self.processBlock(event.Block)
-		case core.ChainSplitEvent:
-			self.processBlock(event.Block)
+		exp--
+		if len(res.prices) > 0 {
+			lps = append(lps, res.prices...)
+		} else {
```

Comment:
Instead of the

```
if len(res.prices) > 0 {
  lps = append(lps, res.prices...)
} else {
```

construct, please use 

```
if len(res.prices) > 0 {
  lps = append(lps, res.prices...)
  continue
}
```

It avoids these deep nestings that you just made here.